### PR TITLE
Show message and reply retrieval progress message

### LINKED ISSUE #1607: Show the progress of message and reply retrieval after a metadata sync
## Description
This idea is closely related to https://github.com/freedomofpress/securedrop-client/issues/1608
The volume of sources and messages at our organization means that metadata syncs and the resulting message and reply download jobs can take quite a long time (around 10-20 minutes). 
As a user it's difficult to know the state of the system during this time - whether or not messages being downloaded and how up to date the sources and messages are. We'd like to be able to give the user explicit feedback about the progress of syncs and resulting download jobs so that this state is more clear. 
We'd like to replace the "Retrieving new messages" status that is displayed after a metadata sync is complete with "Retrieving `n` messages" where `n` is the number of message and reply download jobs queued for processing. Every time jobs complete or are added to the queue, the message would refresh. When all downloads are complete, the message would disappear. 

## How will this impact [SecureDrop users](https://github.com/freedomofpress/securedrop-ux/wiki/Users)?
This change should improve user experience, especially at organizations where sync and download jobs take a significant amount of time.

## How would this affect the SecureDrop Workstation [threat model](https://github.com/freedomofpress/securedrop-workstation/#threat-model)?
This proposal should not have any impact on SecureDrop Workstation's threat model.

## User Stories

As a user of the workstation, I'd like to see a message telling me how many message and reply download jobs are queued so that I know how up-to-date my sources and message are.

### PR DESCRIPTION
**Pull request by zekehuntergreen**
## Description

Fixes #1607.

- Currently after a metadata sync completes, if there are messages and replies to download from the server, the client shows a message "Retrieving messages" for 2.5 seconds before adding a job per message / reply to a queue.
- Due to the Guardian's volume of sources, messages, and replies, processing all message and reply download jobs after a metadata sync has completed can take a long time (around 20 minutes). 
- We would like to provide an easy way for the user to track download progress.
   - Note: as messages are downloaded, the user can see the text being populated (i.e.`<Message not yet available>` being replaced by the decrypted message text) allowing the user to approximate overall download progress.

- This change adds message to the left side of the top nav bar showing the number of message and reply download jobs in the queue when the queue is non-empty, updating every time a job is queued or dequeued (e.g. "Retrieving 23 messages").

### Example

https://user-images.githubusercontent.com/17057932/205993457-551796fa-f1b1-43de-813f-fc303445cc59.mov

## Test Plan

- [ ] Confirm that CI is green :green_apple: 
- [ ] The count of messages is displayed when message download is in progress
- [ ] The count of messages is not displayed once downloading finishes
- [x] Signals are tested without mocking the object under test :dancers: 

We plan to merge https://github.com/guardian/securedrop-client/pull/1 and deploy our fork of the project to production.

## Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations

### CODE REVIEW

**Comment by gonzalo-bulnes** [comment]
Hi @zekehuntergreen! Thanks for opening the draft PR early, it's really nice to see what you're working on. Don't hesitate mentioning me or any of @freedomofpress/securedrop-maintainers in a comment if you need anything. If there is any non-public aspect of the context of your work that you'd like to discuss with us, there are also other channels we can use.

I don't remember seeing Circle CI fail in [this specific way][error] before, but if you want to take advantage of the continuous integration pipeline I think you might need to sign into Circle CI with your GitHub account and authorize it to access your fork.



  [error]: https://app.circleci.com/pipelines/github/freedomofpress/securedrop-client/2895

File: tests/test_queue.py
Comment (gonzalo-bulnes) [inline tests/test_queue.py:513]: :wave: Hello! I've got a special request for you @zekehuntergreen: when testing Qt signal emissions, can you please use `QSignalSpy` instead of `patch`? I know this is not what the rest of the file does, but consistency is not the best choice in this specific case.

Example:
```py
from PyQt5.QtTest import QSignalSpy

# Only if needed, to have an event loop running to process signals:
from tests.helper import app  # noqa: F401


MAX_SIGNAL_WAITING_TIME = 50  # milliseconds are usually more than enough, default is 5 seconds which is a lot to wait for a failure.

message_sent_emissions = QSignalSpy(phone.message_sent)
assert message_sent_emissions.isValid()  # sanity check

message_sent_emissions.wait(MAX_SIGNAL_WAITING_TIME)  # will proceed as soon as the signal is detected
assert len(message_sent_emissions) = 1
assert message_sent_emissions[0] = ["Alice", "Bob", "Hi Bob!"]  # list of arguments of the first message
```

The spy object behaves like a list as far as `len()` and the bracket notation are concerned. You can access the arguments of the calls as if the object was a list of them: `[list_of_args_of_first_message, list_of_args_of_second_message, etc]`.


There are a few advantages to using [`QSignalSpy`][doc]:
- mocking or patching the object under test is less that ideal in general
- `QSignalSpy` allows to work with the actual signals, even though they are asynchronous
- `QSignalSpy` works with re-emitted signals, no so mocking which fails in a non-descriptive way
- it is the recommended way to assert on signal emissions, and I'm pushing for embracing Qt instead of working around it :slightly_smiling_face: 

Of course I'm happy to help doing the refactoring if you want!

  [doc]: https://doc.qt.io/qt-5.15/qsignalspy.html

Comment (zekehuntergreen) [reply tests/test_queue.py]: Thanks so much for the suggestion, @gonzalo-bulnes ! I've replaced the patches with `QSignalSpy` in both tests. That is much neater than patching the signal.

File: securedrop_client/logic.py
Comment (gonzalo-bulnes) [inline securedrop_client/logic.py:418]:
```suggestion
    @pyqtSlot(int)
    def _on_main_queue_updated(self, num_items: int) -> None:
```

Decorating [slots][doc] is [mostly conventional][so] but we're starting to do it systematically. You can import `pyqtSlot` from `PyQt5.QtCore` if it isn't already. The decoration has the nice advantage of expressing the intended use clearly, and helps keeping in mind that nothing too time-consuming should be happening in the method.

  [so]: https://stackoverflow.com/a/40330912
  [doc]: https://doc.qt.io/qt-5.15/signalsandslots.html
  
**Nitpicking**: The method could be private. Since it's new, we know it's not being used anywhere else, and this might be the best time to make that explicit!

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [inline securedrop_client/queue.py:37]:
```suggestion
    def __init__(self, *args: Any, queue_updated_signal: Optional[pyqtSignal] = None, **kwargs: Any):
```
I am surprised that **mypy** didn't catch the implicit `Optional`. :shrug: 


> Explicit is better than implicit. -- [PEP 20][PEP20] - The Zen of Python

  [PEP20]: https://peps.python.org/pep-0020/

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [inline securedrop_client/queue.py:107]:
```suggestion
        queue_updated_signal: Optional[pyqtSignal] = None,
```

File: tests/test_logic.py
Comment (gonzalo-bulnes) [inline tests/test_logic.py:1779]:
```suggestion

    # fake APIJobQueue's emitting main_queue_updated when the job is queued and dequeued
    co.api_job_queue.main_queue_updated.emit(1)
    set_status.assert_called_once_with("Retrieving 1 message", 0)
    
    co.api_job_queue.main_queue_updated.emit(2)
    set_status.assert_called_once_with("Retrieving 2 messages", 0)

```

Might be worth adding  one test case at the boundary between the singular and the plural forms of the message for completeness.

File: tests/test_queue.py
Comment (gonzalo-bulnes) [inline tests/test_queue.py:506]:
```suggestion
    job_queue.enqueue(message_download_job)
    job_queue.enqueue(reply_download_job)
    main_queue_updated_emissions.wait(MAX_SIGNAL_WAITING_TIME)
    main_queue_updated_emissions.wait(MAX_SIGNAL_WAITING_TIME)
```

If enqueuing two jobs, I think it's prudent waiting for two signals. In practice the signals are processed extremely quickly in a small test like this, and granted the double waiting is not the most elegant... but there is a race condition, and waiting double doesn't actually make the test slower.

Truth be told, you might not need to wait _at all_ for the test to pass, but I'd rather treat the signals as if it weren't _almost-synchrounous_ in practice, so that their caveats get printed in our minds when reading the code.

File: tests/test_queue.py
Comment (gonzalo-bulnes) [inline tests/test_queue.py:514]: Same here, I'd wait twice.

File: tests/test_queue.py
Comment (gonzalo-bulnes) [inline tests/test_queue.py:530]: Some indentation is missing in this test and the one above.

```py
    with threads(2) as [main_thread, file_download_thread]:
        job_queue = ApiJobQueue(mock_client, mock_session_maker, main_thread, file_download_thread)

        # Caution: when using the threads() context manager, you need to indent the entire block,
        # because the threads are stopped when exiting the context. (That's what's actually being managed.)
        message_download_job = SeenJob("mock", "mock", "mock")
        # etc.
```

References:

- [Definition of  the `threads` context manager](https://github.com/freedomofpress/securedrop-client/blob/0.8.1/securedrop_client/app.py#L190-L204)
- It was introduced because the Qt really dislikes when `QThread` objects are destroyed before the corresponding threads are closed. (https://github.com/freedomofpress/securedrop-client/commit/5d1dd0c6b84e444822f16680b0cb1b355a44a0a9).

**Review by gonzalo-bulnes — CHANGES_REQUESTED** [review]
I added a couple of items to the **test plan** in the PR description, and left you a few comments. (The little video is a really nice touch BTW.)

There are a couple that you should fix (use of `threads()` and `wait`s) and others for which YMMV, feel free to pick and choose @zekehuntergreen :slightly_smiling_face: 

Otherwise this is a great enhancement, and I'm really glad you contributed it back! I'll try to keep an eye on your fork, but I would almost certainly have missed this one!

_Note: I'll go for the big red scary "Request changes" option, I hope it wasn't that unwelcoming, but oh, well. Please don't read anything but appreciation in it!_

**Comment by gonzalo-bulnes** [comment]
Oh, by the way: the two build failures are unrelated to your changes.

- Rebasing your branch onto the latest `main` will fix the first one (the missing file).
- And we're taking acre of the second will fix the second. Once it's merged the same rebase will  fix it. #1610

File: securedrop_client/queue.py
Comment (zekehuntergreen) [reply securedrop_client/queue.py:37]: After reading the section in the docs on [None and Optional handling](https://mypy.readthedocs.io/en/stable/command_line.html#none-and-optional-handling) this surprised me too! It sounds like the default behavior changed as of version 0.980 so that parameters with `None` as a default value are no longer treated as Optional be default. 

I think what's going on is that mypy can't find a library stub for `PyQt5.QtCore` so in the absence of a type for `pyqtSignal`, it's being treated as `Any`. Since `None` is a subtype of `Any`, mypy doesn't raise an "incompatible default for argument" error.
It looks like there is a [stub library for PyQt5](https://pypi.org/project/PyQt5-stubs/) which it might be worth looking into adding to the project! (I'm happy to raise an issue or PR on this) 
In the meantime I'll use the Optional type here in the interest of making the intended type more clear.

File: tests/test_logic.py
Comment (zekehuntergreen) [reply tests/test_logic.py:1779]: Good call!

File: securedrop_client/logic.py
Comment (zekehuntergreen) [reply securedrop_client/logic.py]: Thanks, I wasn't aware of this decorator!

File: tests/test_queue.py
Comment (zekehuntergreen) [reply tests/test_queue.py]: Thanks very much for the feedback and extra context!

Comment (zekehuntergreen) [reply tests/test_queue.py]: nerdy python pun not intended 🙂

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [inline securedrop_client/queue.py:31]: @cfm Since you've been working on the queues: just leaving a note here that @zekehuntergreen named this class in a consistent manner, and I'm deliberately not poking into the name at this time.

The `Runnable...` class names don't seem very meaningful to me (they seem very jargon and implementation-oriented to start with), but I don't have formed thoughts on this area of the code yet, so I'm not suggesting we change anything at this point. 

Just letting that said so that it's in your radar and for future reference. :slightly_smiling_face:

File: tests/test_logic.py
Comment (gonzalo-bulnes) [inline tests/test_logic.py:1778]: Oh, I missed this. When emitting signals in tests, it is always prudent to set up a spy to wait on them. It is verbose, but again, the signals are processed asynchronously, and nothing guarantees that they'll have been received in in the interval between their emission and the assertion, especially when the assertion statement immediately follows the emission.

```py
# This is probably best placed in the "Setup/Arrange" area of the test, typically towards the beginning?
# But maybe it's better close to where it's used! You've got better context than me to make that decision!
MAX_SIGNAL_WAITING_TIME = 50
main_queue_updated_emissions = QSignalSpy(job_queue.main_queue.queue.queue_updated_signal)

# [...]

# fake APIJobQueue's emitting main_queue_updated when the job is queued and dequeued
co.api_job_queue.main_queue_updated.emit(1)
main_queue_updated_emissions.wait(MAX_SIGNAL_WAITING_TIME)
set_status.assert_called_once_with("Retrieving 1 message", 0)

Comment (gonzalo-bulnes) [reply tests/test_logic.py:1778]: This is true for all three "Act"/"Assert" sequences, of course.

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [reply securedrop_client/queue.py]: Yes, we recently upgraded **mypy** and got that change (https://github.com/freedomofpress/securedrop-client/pull/1599).

I think your hypothesis is very very likely to be correct. It makes a lot of sense to me at least, and I'll give a try to the stub library. Thanks for doing the research @zekehuntergreen! :star2:

**Review by gonzalo-bulnes — CHANGES_REQUESTED** [review]
This is looking great.

1. I missed one signal waiting point in my first review.
2. You can now rebase onto the _latest_ `main` whenever you're ready, and that will fix the failing CI build. 
  For what is worth, I usually do that as a one-liner:
   ```sh
   git checkout main && git pull origin main && git checkout - && git rebase -S main
   # assumes you're currently in the branch you want to rebase, "git checkout -" will come back to it like "cd -" goes back to the directory you were previously in
   # assumes the freedomofpress remote is called "origin", which might not be the case if you've cloned your fork
   # you don't need/want the "-S" rebase option if you're not signing your commits
     ```
3. If you want to squash/group commits, feel free to do so now, because I think we're pretty close! The change set is small enough that the churn of squashing won't make review any more difficult.



----

**Note**: If you want to be extra cautious, you can make a "backup" of your branch before any flavor of `git rebase`:

```sh
git branch show-retrieval-progress-message-backup show-retrieval-progress-message
```

As a personal thing of mine, I offer a lifetime warranty (for real!) to help you recover your git sanity if you made such a "backup" and rebasing ends up in a mess for any reason. :wink:

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [reply securedrop_client/queue.py]: That's it! You're correct.

Adding the **pyqt5-stub** library surfaces a series of type errors and warnings. (Expected.)

I'll open a separate PR for it, as it will likely take a little bit of time to fix them all. But I can confirm that **mypy** notices implicitly optional `pyqtSignal`s now :rocket:

Comment (gonzalo-bulnes) [reply securedrop_client/queue.py]: By the way, the signals we're passing as arguments might actually be of type `pyqtBoundSignal`! :bulb:

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [inline securedrop_client/queue.py:38]: Given https://github.com/freedomofpress/securedrop-client/pull/1604#discussion_r1047896459, this should be `pyqtBoundSignal`. (We're never manipulating unbound signals that I know of, they're always bound to an instance of the class where we define them.)

The type can be imported from `PyQt5.QtCore`.

```suggestion
        self, *args: Any, queue_updated_signal: Optional[pyqtBoundSignal] = None, **kwargs: Any
```

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [inline securedrop_client/queue.py:109]:
```suggestion
        queue_updated_signal: Optional[pyqtBoundSignal] = None,
```

File: securedrop_client/queue.py
Comment (gonzalo-bulnes) [reply securedrop_client/queue.py:38]: Don't worry about other occurrences that you didn't introduce, we'll fix all those at once in #1611.

**Review by gonzalo-bulnes — CHANGES_REQUESTED** [review]
Breaking news: there is a specific type for `pyqtBoundSignal`s, and it happens those are the ones we use/need hints for. Your finding has beneficial secondary effects @zekehuntergreen! TIL :bulb:

File: tests/test_logic.py
Comment (zekehuntergreen) [reply tests/test_logic.py:1778]: Thanks, I missed this one as well!

File: securedrop_client/queue.py
Comment (zekehuntergreen) [reply securedrop_client/queue.py]: Amazing, I'll take a look at #1611

**Comment by zekehuntergreen** [comment]
@gonzalo-bulnes Thanks so much for the incredibly thorough and helpful review!
I've rebased onto the latest main and checks seem to be passing 🙌 
And thanks for the very generous offer to help me escape from git hell after a potential botched rebase! I'm not used to squashing commits (other than squashing and merging a pull request) but seem to have gotten away with squashing and renaming without any major issues 😅

**Comment by gonzalo-bulnes** [comment]
The commit grouping you chose reads great! :clap: I'm glad it went well, manipulating branches via `git rebase` or `git rebase --interactive` becomes less prone to surprises with practice.

_If that helps, it helped me to think of rebase operations as applying changes (commits) in a different contexts (with this [definition of context][ctx] - I never wrote the continuation, but the gist of it follows :point_right:). From there, conflicts are differences of context that happen to be significant to the changes being applied (that is, to achieve the same goal, the exact recipe/commit/set of differences might have to change if the context changes, even if some context changes are jut not significant to any given recipe/commit). The more the original and new context differ, the more likely any given commit will have to be modified to achieve what it was intended to achieve, and usually, you need to pick a little bit of both sides of the conflict - otherwise there wouldn't be a conflict. Typically, swapping commits around is more likely to cause conflicts, because one was part of the _context_ of the other, and unless your commits are very modular, the second commit might have been building upon the first one._

_A little bit counter-intuitively, conflicts are actually the good case, when significant differences in context can be detected automatically by Git. If those differences happen in distant places of any given file, they'll be silent, and you have to rely on your test suite to catch them, a less good case that sometimes happens._

  [ctx]: https://gonzalo-bulnes.github.io/blog/gardening_with_git/context-from-scratch.html

**Review by gonzalo-bulnes — APPROVED** [review]
## Test plan

- [x] Confirm that CI is green :green_apple: 
- [x] The count of messages is displayed when message download is in progress
- [x] The count of messages is not displayed once downloading finishes
- [x] The UI looks like the video
- [x] Signals are tested without mocking the object under test :dancers:

**Comment by gonzalo-bulnes** [comment]
Thank you again @zekehuntergreen :tada: :rocket:

**Pull request merged**
